lib/messageReceiver.ts

```diff
@@ -0,0 +1,232 @@
+ * @class MessageReceiver
+ */
+export class MessageReceiver extends LinkEntity {
+  /**
+   * @property {string} receiverType The type of receiver: "batching" or "streaming".
+   */
+  receiverType: ReceiverType;
+  /**
+   * @property {number} [maxConcurrentCalls] The maximum number of messages that should be
+   * processed concurrently while in peek lock mode. Once this limit has been reached, more
+   * messages will not be received until messages currently being processed have been settled.
+   * Default: 1
+   */
+  maxConcurrentCalls?: number;
+  /**
+   * @property {number} [receiveMode] The mode in which messages should be received.
+   * Default: ReceiveMode.peekLock
+   */
+  receiveMode: ReceiveMode;
+  /**
+   * @property {boolean} [autoComplete] Indicates whether `Message.complete()` should be called
+   * automatically after the message processing is complete while receiving messages with handlers
+   * or while messages are received using receiveBatch(). Default: false.
+   */
+  autoComplete: boolean;
+  /**
+   * @property {Receiver} [_receiver] The AMQP receiver link.
+   * @protected
+   */
+  protected _receiver?: Receiver;
+  /**
+   * @property {OnMessage} _onMessage The message handler provided by the user that will be wrapped
+   * inside _onAmqpMessage.
+   * @protected
+   */
+  protected _onMessage?: OnMessage;
+  /**
+   * @property {OnMessage} _onMessage The error handler provided by the user that will be wrapped
+   * inside _onAmqpError.
+   * @protected
+   */
+  protected _onError?: OnError;
+  /**
+   * @property {OnMessage} _onMessage The message handler that will be set as the handler on the
+   * underlying rhea receiver for the "message" event.
+   * @protected
+   */
+  protected _onAmqpMessage: OnAmqpEvent;
+  /**
+   * @property {OnMessage} _onMessage The message handler that will be set as the handler on the
+   * underlying rhea receiver for the "receiver_error" event.
+   * @protected
+   */
+  protected _onAmqpError: OnAmqpEvent;
+
+  constructor(context: ClientEntityContext, receiverType: ReceiverType, options?: ReceiveOptions) {
+    super(`${context.entityPath}`, context);
+    if (!options) options = {};
+    this.receiverType = receiverType;
+    this.address = `${this._context.entityPath}`;
+    this.audience = `${this._context.namespace.config.endpoint}${this._context.entityPath}`;
+    this.maxConcurrentCalls = options.maxConcurrentCalls != undefined ?
+      options.maxConcurrentCalls : 1;
+    this.autoComplete = options.autoComplete != undefined ? options.autoComplete : false;
+    this.receiveMode = options.receiveMode || ReceiveMode.peekLock;
+    this._onAmqpMessage = async (context: EventContext) => {
+      const bMessage = new Message(context.message!, context.delivery!);
+      bMessage.body = this._context.namespace.dataTransformer.decode(context.message!.body);
+      try {
+        await this._onMessage!(bMessage);
```

Comment:
having the `!` here is odd because this is depending on something that extends it filling in the implementation for it. if you want to make the assertion it probably makes more sense to do it on the declaration with `protected _onMessage!: OnMessage;`

Reply:
i'd actually recommend having the `StreamingReceiver` and `BatchingReceiver` contain a `MessageReceiver` rather than extending it so you can just pass the necessary things in the constructor. that way you don't have to have these implicit contracts where one thing sets a property on the other